quic/s2n-quic-core/src/recovery/bbr.rs

```diff
@@ -503,6 +594,87 @@ impl BbrCongestionController {
         (MIN_PIPE_CWND_PACKETS * self.max_datagram_size) as u32
     }
 
+    /// Updates the congestion window based on the latest model
+    fn set_cwnd(&mut self, newly_acked: usize) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.4.6
+        //# BBRSetCwnd():
+        //#   BBRUpdateMaxInflight()
+        //#   BBRModulateCwndForRecovery()
+        //#   if (!BBR.packet_conservation) {
+        //#     if (BBR.filled_pipe)
+        //#       cwnd = min(cwnd + rs.newly_acked, BBR.max_inflight)
+        //#     else if (cwnd < BBR.max_inflight || C.delivered < InitialCwnd)
+        //#       cwnd = cwnd + rs.newly_acked
+        //#     cwnd = max(cwnd, BBRMinPipeCwnd)
+        //#  }
+        //#  BBRBoundCwndForProbeRTT()
+        //#  BBRBoundCwndForModel()
+
+        // From BBRModulateCwndForRecovery()
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.4.4
+        //#   if (BBR.packet_conservation)
+        //#     cwnd = max(cwnd, packets_in_flight + rs.newly_acked)
+
+        let max_inflight = self.max_inflight().try_into().unwrap_or(u32::MAX);
+        let initial_cwnd = Self::initial_window(self.max_datagram_size);
+        let mut cwnd = self.cwnd;
+
+        if self.recovery_state.packet_conservation() {
+            cwnd = cwnd.max(self.bytes_in_flight.saturating_add(newly_acked as u32));
+        } else if self.full_pipe_estimator.filled_pipe() {
+            cwnd = (cwnd + newly_acked as u32).min(max_inflight);
+        } else if cwnd < max_inflight || self.bw_estimator.delivered_bytes() < initial_cwnd as u64 {
+            cwnd += newly_acked as u32;
+        }
+
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.4.5
+        //# BBRBoundCwndForProbeRTT():
+        //#   if (BBR.state == ProbeRTT)
+        //#     cwnd = min(cwnd, BBRProbeRTTCwnd())
+        if self.state.is_probing_rtt() {
+            cwnd = cwnd.min(self.probe_rtt_cwnd());
+        }
+        self.cwnd = cwnd.clamp(self.minimum_window(), self.bound_cwnd_for_model());
+    }
+
+    /// Returns the maximum congestion window bound by recent congestion
+    fn bound_cwnd_for_model(&self) -> u32 {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.4.7
+        //# BBRBoundCwndForModel():
+        //#   cap = Infinity
+        //#   if (IsInAProbeBWState() and
+        //#       BBR.state != ProbeBW_CRUISE)
+        //#     cap = BBR.inflight_hi
+        //#   else if (BBR.state == ProbeRTT or
+        //#            BBR.state == ProbeBW_CRUISE)
+        //#     cap = BBRInflightWithHeadroom()
+        //#
+        //#   /* apply inflight_lo (possibly infinite): */
+        //#   cap = min(cap, BBR.inflight_lo)
+        //#   cap = max(cap, BBRMinPipeCwnd)
+        //#   cwnd = min(cwnd, cap)
```

Comment:
Do we need this operation as well?

Reply:
`BBRBoundCwndForModel` in the pseudocode mutates the cwnd, whereas I just return the cap and then apply the clamping on the result in `set_cwnd()`. I didn't really like having a bunch of different methods mutating the congestion window

Reply:
Makes sense